libsolidity/formal/PredicateSort.cpp

```diff
@@ -60,7 +60,7 @@ SortPointer constructorSort(ContractDefinition const& _contract, SymbolicState&
 		return functionSort(*constructor, &_contract, _state);
 
 	return make_shared<FunctionSort>(
-		vector<SortPointer>{_state.errorFlagSort(), _state.thisAddressSort(), _state.stateSort()} + stateSorts(_contract),
+		vector<SortPointer>{_state.errorFlagSort(), _state.thisAddressSort(), _state.txSort(), _state.stateSort()} + stateSorts(_contract),
```

Comment:
I think this does not match the docs in `PredicateSort.h`:
```
 * 4. Constructor entry/summary
 * The summary of an implicit constructor. Signature:
 * constructor_summary(error, this, txData, blockchainState, blockchainState', stateVariables').
```
The sort corresponding to either `blockchainState` or `blockchainState'` seems to be missing.

Reply:
Nice catch! I think the predicate should indeed have both, since the state might have changed between the beginning and end of the constructor. So the spec is correct. I might push a separate PR for this, since it's unrelated to this PR.

Reply:
Fixed in https://github.com/ethereum/solidity/pull/10014

Reply:
> state might have changed between the beginning and end of the constructor

Now that I think about it - can the state really change? You cannot make external calls from a constructor and even if you could, the `msg.*`, `tx.*` and `block.*` properties return back to original values when the external call returns. No external calls means that balances stay the same. The only thing I can think of that could change is `gasleft()`.

Reply:
Why can't you make external calls from a constructor? The following works:
```
abstract contract D {
    function f() external virtual;
}
contract C {
  constructor(D d) public payable {
      address payable a = address(2);
      a.transfer(2 ether);
      d.f();
  }
}
```
Also, as the example above, the balance of the analyzed contract itself might change.
I agree that the tx data (msg, tx, block) stays the same, that's why there's only one instance of `tx` in that predicate, but 2 of state because `balances` might change

Reply:
Ah, ok. To be honest, I never really tried it but I remember reading somewhere that external calls in constructor just revert.

Reply:
That might actually have been about calling contract's own external methods. In that case yeah, the state can change.